dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/TimeFieldSqlRenderer.java

```diff
@@ -0,0 +1,146 @@
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.analytics.event.data;
+
+import static org.hisp.dhis.analytics.EventOutputType.ENROLLMENT;
+import static org.hisp.dhis.analytics.TimeField.EVENT_DATE;
+import static org.hisp.dhis.analytics.event.data.JdbcEventAnalyticsManager.OPEN_IN;
+import static org.hisp.dhis.analytics.util.AnalyticsSqlUtils.ANALYTICS_TBL_ALIAS;
+import static org.hisp.dhis.analytics.util.AnalyticsSqlUtils.DATE_PERIOD_STRUCT_ALIAS;
+import static org.hisp.dhis.analytics.util.AnalyticsSqlUtils.quote;
+import static org.hisp.dhis.analytics.util.AnalyticsSqlUtils.quoteAlias;
+import static org.hisp.dhis.common.DimensionalObject.PERIOD_DIM_ID;
+import static org.hisp.dhis.common.IdentifiableObjectUtils.getUids;
+import static org.hisp.dhis.commons.util.TextUtils.getQuotedCommaDelimitedString;
+import static org.hisp.dhis.util.DateUtils.getMediumDateString;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import lombok.RequiredArgsConstructor;
+
+import org.hisp.dhis.analytics.TimeField;
+import org.hisp.dhis.analytics.event.EventQueryParams;
+import org.hisp.dhis.common.DimensionalItemObject;
+import org.hisp.dhis.jdbc.StatementBuilder;
+import org.hisp.dhis.period.Period;
+import org.hisp.dhis.program.AnalyticsPeriodBoundary;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+class TimeFieldSqlRenderer
+{
+
+    private final StatementBuilder statementBuilder;
+
+    public String renderTimeFieldSql( EventQueryParams params, Collection<TimeField> allowedTimeFields )
+    {
+        StringBuilder sql = new StringBuilder();
+
+        Optional<TimeField> timeField = TimeField.of( params.getTimeField() )
+            .filter( allowedTimeFields::contains );
+
+        if ( params.hasNonDefaultBoundaries() )
+        {
+            for ( AnalyticsPeriodBoundary boundary : params.getProgramIndicator().getAnalyticsPeriodBoundaries() )
+            {
+                sql
+                    .append( statementBuilder.getBoundaryCondition( boundary, params.getProgramIndicator(),
+                        params.getTimeFieldAsField(), params.getEarliestStartDate(), params.getLatestEndDate() ) )
+                    .append( " " );
+            }
+        }
+        else if ( params.hasStartEndDate() )
+        {
+
+            String timeCol;
+            if ( ENROLLMENT.equals( params.getOutputType() ) )
+            {
+                timeCol = quoteAlias( TimeField.ENROLLMENT_DATE.getField() );
```

Comment:
Should last updated aggregation should also be supported for enrollment type indicators?